src/shared/codelens/localLambdaRunner.ts

```diff
@@ -0,0 +1,357 @@
+import { SamTemplateGenerator } from '../templates/sam/samTemplateGenerator'
+import { ExtensionDisposableFiles } from '../utilities/disposableFiles'
+
+interface LambdaLocalInvokeArguments {
+    document: vscode.TextDocument,
+    range: vscode.Range,
+    handlerName: string,
+    debug: boolean,
+    workspaceFolder: vscode.WorkspaceFolder
+}
+
+interface SAMTemplateEnvironmentVariables {
+    [resource: string]: {
+        [key: string]: string
+    }
+}
+
+export class LocalLambdaRunner {
+
+    private static readonly TEMPLATE_RESOURCE_NAME: string = 'awsToolkitSamLocalResource'
+    private static readonly SAM_LOCAL_PORT_CHECK_RETRY_INTERVAL_MILLIS: number = 125
+    private static readonly SAM_LOCAL_PORT_CHECK_RETRY_TIMEOUT_MILLIS_DEFAULT: number = 30000
+
+    private _baseBuildFolder?: string
+    private readonly _debugPort?: number
+
+    public constructor(
+        private readonly configuration: SettingsConfiguration,
+        private readonly localInvokeArgs: LambdaLocalInvokeArguments,
+        debugPort: number | undefined,
+        private readonly runtime: string,
+        private readonly outputChannel: vscode.OutputChannel,
+        private readonly processInvoker: SamCliProcessInvoker,
+        private readonly taskInvoker: SamCliTaskInvoker
+    ) {
+        if (localInvokeArgs.debug && !debugPort) {
+            throw new Error('Debug port must be provided when launching in debug mode')
+        }
+
+        this._debugPort = debugPort
+    }
+
+    public async run(): Promise<void> {
+        try {
+            this.outputChannel.show(true)
+            this.outputChannel.appendLine(
+                localize(
+                    'AWS.output.sam.local.start',
+                    'Preparing to run {0} locally...',
+                    this.localInvokeArgs.handlerName
+                )
+            )
+
+            const rootCodeFolder: string = await this.determineRootCodeFolder()
+
+            const inputTemplate: string = await this.generateInputTemplate(rootCodeFolder)
+            const samBuildTemplate: string = await this.executeSamBuild(rootCodeFolder, inputTemplate)
+
+            await this.invokeLambdaFunction(samBuildTemplate)
+
+        } catch (err) {
+            console.log(err)
+            const error = err as Error
+
+            this.outputChannel.appendLine(
+                localize(
+                    'AWS.output.sam.local.error',
+                    'Error: {0}',
+                    error.message
+                )
+            )
+
+            vscode.window.showErrorMessage(
+                localize(
+                    'AWS.error.during.sam.local',
+                    'An error occurred trying to run SAM Application locally: {0}',
+                    error.message
+                )
+            )
+
+            return
+        }
+
+    }
+
+    public get debugPort(): number {
+        if (!this._debugPort) {
+            throw new Error('Debug port was expected but is undefined')
```

Comment:
Nit: I feel like this should be the responsibility of the caller. For example, a caller might want to check whether `debugPort` is defined.

Reply:
Understood, improvements are needed. The purpose of the CR is to move the code without any implementation changes.